include/aikido/robot/ros/RosRobot.hpp

```diff
@@ -18,21 +21,198 @@ AIKIDO_DECLARE_POINTERS(RosRobot)
+  /// the inactive list. \param[in] controllerMode The corresponding control
+  /// mode to add to the inactive list. \return A robot-unique non-negative ID
+  /// (negative implies failure)
+  std::string registerExecutor(
+      aikido::control::ExecutorPtr executor,
+      std::string controllerName,
+      hardware_interface::JointCommandModes controllerMode,
+      std::string desiredName);
+
+  ///
+  /// Loads the controller. Throws a runtime_error if controller cannot be
+  /// loaded. Add an (executor, controller) to the inactive (executor,
+  /// controller, control mode) list Operations on this executor do not affect
+  /// the control mode. Releases DoFs held by executor if held.
+  ///
+  /// \param[in] executor The Executor to add to the inactive list.
+  /// \param[in] controllerName The corresponding controller to load and add to
+  /// the inactive list. \return A robot-unique non-negative ID (negative
+  /// implies failure)
+  std::string registerExecutor(
+      aikido::control::ExecutorPtr executor, 
+      std::string controllerName,
+      std::string desiredName);
+
+  ///
+  /// Add an executor to the inactive executors list.
+  /// Operations on this executor do not affect the controller or control mode.
+  /// Releases DoFs held by executor if held.
+  ///
+  /// \param[in] executor The Executor to add to the inactive list.
+  /// \return A robot-unique non-negative ID (negative implies failure)
+  std::string registerExecutor(aikido::control::ExecutorPtr executor, std::string desiredName = "") override;
+
+  ///
+  /// Deactivates the current active executor.
+  /// Starts the corresponding controller and switches to corresponding control
+  /// mode. Sets an executor from the inactive executor list to be the active
+  /// executor. Holds and releases DoFs as needed.
+  ///
+  /// \param[in] id of executor on executor list
+  /// \return True if successful. If false, all executors are inactive.
+  bool activateExecutor(std::string id) override; 
+
+  using Robot::activateExecutor; // This un-hides Robot::activateExecutor(const aikido::control::ExecutorType type)
+
+  ///
+  /// Sets the ros controller service client which enables controller loading.
+  ///
+  /// \param[in] rosControllerServiceClient ros controller service client.
+  void setRosLoadControllerServiceClient(
+      const std::shared_ptr<::ros::ServiceClient>& rosLoadControllerServiceClient); 
+
+  ///
+  /// Sets the ros controller service client which enables controller switching.
+  ///
+  /// \param[in] rosControllerServiceClient ros controller service client.
+  void setRosSwitchControllerServiceClient(
+      const std::shared_ptr<::ros::ServiceClient>& rosSwitchControllerServiceClient);
+
+  ///
+  /// Sets the ros joint mode command client which enables controller mode switching.
+  ///
+  /// \param[in] rosJointModeCommandClient ros joint mode command client.
+  void setRosJointModeCommandClient(
```

Comment:
Similarly, I would not have an external class own the joint mode clients.

Have this function be `setRosJointModeClient(string serverName, vector joints = empty, other args = default)` and initialize the client internally. If not provided, use the list of joint names from the robot metaskeleton.